src/doorman/Channel.cpp

```diff
@@ -0,0 +1,393 @@
+    byte buf[] = {IAC, a, b};
+    fd.write(buf);
+}
+
+static void SendOpt(const Fd &fd, byte a) {
+    byte buf[] = {IAC, SB, a, TELQUAL_SEND, IAC, SE};
+    fd.write(buf);
+}
+
+void Channel::async_lookup_process(sockaddr_in address, Fd reply_fd) {
+    /*
+     * Firstly, and quite importantly, close all the descriptors we
+     * don't need, so we don't keep open sockets past their useful
+     * life.  This was a bug which meant ppl with firewalled auth
+     * ports could keep other ppls connections open (nonresponding)
+     */
+    doorman_->for_each_channel([](Channel &channel) { channel.close(); });
+
+    // Bail out on a PIPE - this means our parent has crashed
+    signal(SIGPIPE, [](int) {
+        log_out("hostname lookup process exiting on sigpipe");
+        exit(1);
+    });
+
+    auto *ent = gethostbyaddr((char *)&address.sin_addr, sizeof(address.sin_addr), AF_INET);
+
+    std::string hostName = ent ? ent->h_name : inet_ntoa(address.sin_addr);
+    size_t nBytes = hostName.size() + 1;
+    try {
+        reply_fd.write(nBytes);
+        reply_fd.write(hostName.c_str(), nBytes);
+    } catch (const std::runtime_error &re) {
+        log_out("ID: Unable to write to doorman - perhaps it crashed: %s", re.what());
+        return;
+    }
+    // Log the source hostname but mask it for privacy.
+    log_out("ID: Looked up %s", get_masked_hostname(hostName).c_str());
+}
+
+static bool SupportsAnsi(const char *src) {
+    static const char *terms[] = {"xterm",      "xterm-colour", "xterm-color", "ansi",
+                                  "xterm-ansi", "vt100",        "vt102",       "vt220"};
+    size_t i;
+    for (i = 0; i < (sizeof(terms) / sizeof(terms[0])); ++i)
+        if (!strcasecmp(src, terms[i]))
+            return true;
+    return false;
+}
+
+void Channel::closeConnection() {
+    // Log the source IP but masked for privacy.
+    log_out("[%d] Closing connection to %s", id_, get_masked_hostname(hostname_).c_str());
+    close();
+}
+
+void Channel::set_echo(bool echo) {
+    SendCom(fd_, echo ? WONT : WILL, TELOPT_ECHO);
+    echoing_ = echo;
+}
+
+void Channel::on_reconnect_attempt() {
+    if (!fd_.is_open())
+        return;
+    if (!sent_reconnect_message_) {
+        send_to_client("Attempting to reconnect to Xania"sv);
+        sent_reconnect_message_ = true;
+    }
+    send_to_client(".", 1);
+}
+
+void Channel::close() {
+    fd_.close();
+    host_lookup_fd_.close();
+    if (host_lookup_pid_) {
+        if (::kill(host_lookup_pid_, 9) < 0)
+            log_out("Couldn't kill child process (ident process has already exited)");
+        host_lookup_pid_ = 0;
+    }
+}
+
+void Channel::newConnection(Fd fd, sockaddr_in address) {
+    // One day I will be a constructor...
+    this->fd_ = std::move(fd);
+    hostname_ = inet_ntoa(address.sin_addr);
+    port_ = ntohs(address.sin_port);
+    netaddr_ = ntohl(address.sin_addr.s_addr);
+    buffer_.clear();
+    width_ = 80;
```

Comment:
Would it not make sense to have 'int width_  = 80;' in the .h file, rather than default-initing it there, and then overwriting it here?

Reply:
Agreed. The way the code is phrased the Channels are statically allocated and then reused over and over again instead of being destructed and re-constructed. That's a holdover from the way Channel IDs work and the way the code was written, and one I plan on changing. Ideally the Channel would be constructed when it has a connection come in, that way all these variables are initialsed there. I hate the fact we overwrite them here. For the constants, having them in the header file (presumably you mean member initialisation here? (`int width_{80};` etc) makes sense then.

Reply:
Noted in the PR TODOs